tests/v3_api/test_auth.py

```diff
@@ -219,6 +252,98 @@ def validate_add_users_and_groups_to_cluster_or_project(
+    allowed_principal_ids = []
+    for user in auth_setup_data["allowed_users"]:
+        allowed_principal_ids.append(principal_lookup(user, admin_token))
+    allowed_principal_ids.append(principal_lookup(admin_user, admin_token))
+
+    # Add users in allowed list to access rancher-server
+    add_users_to_siteAccess(admin_token, access_mode, allowed_principal_ids)
+
+    for user in auth_setup_data["allowed_users"]:
+        token = login(user, PASSWORD)
+        if AUTH_PROVIDER == "activeDirectory":
+            disable_ad(user, token, expected_status=403)
+            enable_ad(user, token, expected_status=403)
+        if AUTH_PROVIDER == "openLdap":
+            disable_openldap(user, token, expected_status=setup["permission_denied_code"])
+            enable_openldap(user, token, expected_status=setup["permission_denied_code"])
+        if AUTH_PROVIDER == "freeIpa":
+            disable_freeipa(user, token, expected_status=setup["permission_denied_code"])
+            enable_freeipa(user, token, expected_status=setup["permission_denied_code"])
+
+
+def validate_access_control_disable_and_enable_nestedgroups(access_mode):
+    delete_project_users()
+    delete_cluster_users()
+
+    auth_setup_data = setup["auth_setup_data"]
+    admin_user = auth_setup_data["admin_user"]
+    token = login(admin_user, PASSWORD)
+    if AUTH_PROVIDER == "activeDirectory":
+        enable_ad(admin_user, token)
+    if AUTH_PROVIDER == "openLdap":
+        enable_openldap(admin_user, token)
+    if AUTH_PROVIDER == "freeIpa":
+        enable_freeipa(admin_user, token)
+
+    allowed_principal_ids = []
+    for group in auth_setup_data["allowed_nestedgroups"]:
+        allowed_principal_ids.append(principal_lookup(group, token))
+
+    allowed_principal_ids.append(principal_lookup(admin_user, token))
+
+    # Add users in allowed list to access rancher-server
+    add_users_to_siteAccess(token, access_mode, allowed_principal_ids)
+
+    for group in auth_setup_data["allowed_nestedgroups"]:
+        for user in auth_setup_data[group]:
+            login(user, PASSWORD)
+
+    if AUTH_PROVIDER == "activeDirectory" or AUTH_PROVIDER == "openLdap":
+        for user in auth_setup_data["users_under_nestedgroups"]:
+            login(user, PASSWORD, expected_status=setup["permission_denied_code"])
+
+    # Enable nestedgroup feature, so users under nestedgroups can login successfully
+    if AUTH_PROVIDER == "activeDirectory":
+        enable_ad_nestedgroups(admin_user, token)
+    if AUTH_PROVIDER == "openLda[":
+        enable_openldap_nestedgroup(admin_user, token)
+
+    allowed_principal_ids = []
+    for group in auth_setup_data["allowed_nestedgroups"]:
+       allowed_principal_ids.append(principal_lookup(group, token))
+    allowed_principal_ids.append(principal_lookup(admin_user, token))
+
+    # Add users in allowed list to access rancher-server
+    add_users_to_siteAccess(token, access_mode, allowed_principal_ids)
```

Comment:
Isnt this part of code (line 331-337) already executed in lines 308-315? Why are we invoking it again?

Reply:
In case of FreeIPA this part will get executed twice for no reason?